File: c++/design-twitter.cc
class Twitter {
private:
  unordered_map<int, vector<pair<int, int>>> tweets;
  unordered_map<int, unordered_set<int>> follows;
  int inc = 0;

  struct Compare {
    bool operator()(tuple<int, int, int>& a, tuple<int, int, int>& b) {
      return get<0>(a) < get<0>(b);
    }
  };
public:
  Twitter() {}
  
  void postTweet(int userId, int tweetId) {
    tweets[userId].push_back({tweetId, inc++});
  }
  
  vector<int> getNewsFeed(int userId) {
    priority_queue<tuple<int, int, int>, vector<tuple<int, int, int>>, Compare> maxHeap;
    vector<int> topTenTweets;

    if (!tweets[userId].empty()) maxHeap.push({tweets[userId].back().second, userId, tweets[userId].size() - 1});
    for (int followee : follows[userId]) {
      if (tweets[followee].empty()) continue;
      maxHeap.push({tweets[followee].back().second, followee, tweets[followee].size() - 1});
    }

    while (topTenTweets.size() < 10 && !maxHeap.empty()) {
      auto [tweet, poster, index] = maxHeap.top();
      maxHeap.pop();
      topTenTweets.push_back(tweets[poster][index].first);
      if (index) {
        maxHeap.push({tweets[poster][index - 1].second, poster, index - 1});
      }
    }

    return topTenTweets;
  }
  
  void follow(int followerId, int followeeId) {
   follows[followerId].insert(followeeId); 
  }
  
  void unfollow(int followerId, int followeeId) {
    if (follows[followerId].contains(followeeId)) {
      follows[followerId].erase(followeeId);
    }
  }
};

File: python/design-twitter.py
class Twitter:

    def __init__(self):
        self.subscribedTo = defaultdict(set[int])
        self.tweets = defaultdict(list[tuple[int, int]])
        self.time = 0

    def postTweet(self, userId: int, tweetId: int) -> None:
        self.tweets[userId].append((self.time, tweetId))
        self.time += 1
        self.subscribedTo[userId].add(userId)

    def getNewsFeed(self, userId: int) -> List[int]:
        minHeap = []
        for poster in self.subscribedTo[userId]:
            for tweet in self.tweets[poster]:
                if len(minHeap) < 10 or minHeap[0][0] < tweet[0]:
                    heappush(minHeap, (tweet[0], tweet[1]))
                if len(minHeap) > 10:
                    heappop(minHeap)

        minHeap.sort(key=lambda x: x[0], reverse=True)
        return [elem[1] for elem in minHeap]

    def follow(self, followerId: int, followeeId: int) -> None:
        if followerId == followeeId:
            return
        self.subscribedTo[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        if followeeId == followerId:
            return
        if followeeId in self.subscribedTo[followerId]:
            self.subscribedTo[followerId].remove(followeeId)


# Your Twitter object will be instantiated and called as such:
# obj = Twitter()
# obj.postTweet(userId,tweetId)
# param_2 = obj.getNewsFeed(userId)
# obj.follow(followerId,followeeId)
# obj.unfollow(followerId,followeeId)
